### PR TITLE
CodeView: use D-Bus activation for starting up the gnome-builder

### PR DESCRIPTION
**Pull request by krnowak**
https://phabricator.endlessm.com/T14628

### CODE REVIEW

File: js/ui/codingManager.js [hunk 1]
```diff
@@ -267,7 +267,23 @@ const CodingManager = new Lang.Class({
             this._watchdogId = Mainloop.timeout_add(WATCHDOG_TIME,
                                                     Lang.bind(this, this._watchdogTimeout));
 
-            Util.trySpawn(['flatpak', 'run', 'org.gnome.Builder', '-s', '-m', constructCommand(appManifest)]);
```
Comment (erikos) [inline js/ui/codingManager.js]: const Util = imports.misc.util;

We can remove that import then.

**Comment by erikos** [comment]
Ok, so just had a quick test as well, works fine. So, a separate function would be nice as Sam suggested and the removal of the unused util import.

File: js/ui/codingManager.js [hunk 2]
```diff
@@ -267,7 +267,23 @@ const CodingManager = new Lang.Class({
...
+            let params = new GLib.Variant('(sava{sv})', ['load-flatpak', [new GLib.Variant('s', constructCommand(appManifest))], {}]);
+            Gio.DBus.session.call('org.gnome.Builder',
+                                  '/org/gnome/Builder',
+                                  'org.gtk.Actions',
+                                  'Activate',
+                                  params,
+                                  null,
+                                  Gio.DBusCallFlags.NONE,
+                                  -1,
```
Comment (cosimoc) [inline js/ui/codingManager.js]: You should pass `G_MAXINT` here to avoid timeouts

File: js/ui/codingManager.js [hunk 3]
```diff
@@ -267,7 +267,23 @@ const CodingManager = new Lang.Class({
...
+                                  null,
+                                  function (conn, result) {
+                                      try {
+                                          conn.call_finish(result);
+                                      } catch (e) {
+                                          log('Failed to start gnome-builder: ' + e.toString());
```
Comment (cosimoc) [inline js/ui/codingManager.js]: Use `logError(e, 'message')`

File: js/ui/codingManager.js (hunk 2 above)
Comment (krnowak) [inline js/ui/codingManager.js]: Finding javascript's counterpart of the C `G_MAXINT` macro was a bit challenging…

File: js/ui/codingManager.js (hunk 3 above)
Comment (krnowak) [inline js/ui/codingManager.js]: Done.

**Comment by krnowak** [comment]
Refactored to a separate function, but I haven't tested it - somehow the button for rotating the app stopped showing up in the gnome weather application…

**Comment by erikos** [comment]
Thanks Krzesimir. Looking good and tested here as working. I would keep the helper function to construct the command though.
```

diff --git a/js/ui/codingManager.js b/js/ui/codingManager.js
index d923eef..a521719 100644
--- a/js/ui/codingManager.js
+++ b/js/ui/codingManager.js
@@ -246,10 +246,13 @@ const CodingManager = new Lang.Class({
         }
     },
 
-    _startBuilder: function(manifestPath, flatpakId) {
-        // add an app_id_override to the manifest to load
-        let flagValue = manifestPath + '+' + flatpakId + '.Coding';
-        let params = new GLib.Variant('(sava{sv})', ['load-flatpak', [new GLib.Variant('s', flagValue)], {}]);
+    _startBuilder: function(appManifest, flatpakId) {
+        function constructCommand(appManifest) {
+            // add an app_id_override to the manifest to load
+            return appManifest.get_path() + '+' + flatpakId + '.Coding';
+        }
+
+        let params = new GLib.Variant('(sava{sv})', ['load-flatpak', [new GLib.Variant('s', constructCommand(appManifest))], {}]);
         Gio.DBus.session.call('org.gnome.Builder',
                               '/org/gnome/Builder',
                               'org.gtk.Actions',
@@ -283,7 +286,7 @@ const CodingManager = new Lang.Class({
             this._watchdogId = Mainloop.timeout_add(WATCHDOG_TIME,
                                                     Lang.bind(this, this._watchdogTimeout));
 
-            this._startBuilder(appManifest.get_path(),
+            this._startBuilder(appManifest,
                                session.actorApp.meta_window.get_flatpak_id());
             animateBounce(session.buttonApp);
         } else {

```

**Pull request merged**